### PR TITLE
Feature: Town name filtering

### PR DESCRIPTION
**Pull request by telk5093**
It's my first OpenTTD PR, and please let you know I'm very **new to c++ and openttd's codes.**
I referenced this code from town_gui.cpp and other's PR and it works, but very strangely.
So please advice your enhanced code, and you can ignore/edit this commit.

This allows filtering town names like below:
image

### CODE REVIEW

File: src/widgets/town_widget.h [hunk 1]
```diff
@@ -16,6 +16,7 @@
 enum TownDirectoryWidgets {
 	WID_TD_SORT_ORDER,       ///< Direction of sort dropdown.
 	WID_TD_SORT_CRITERIA,    ///< Criteria of sort dropdown.
+	WID_SL_FILTER,           ///< Filter of name.
```
Comment (nielsmh) [inline src/widgets/town_widget.h]: Wrong identifier name, make sure it's unique for each window. `WID_TD_FILTER` instead.

File: src/town_gui.cpp [hunk 2]
```diff
@@ -618,7 +619,8 @@ static const NWidgetPart _nested_town_directory_widgets[] = {
 			NWidget(NWID_HORIZONTAL),
 				NWidget(WWT_TEXTBTN, COLOUR_BROWN, WID_TD_SORT_ORDER), SetDataTip(STR_BUTTON_SORT_BY, STR_TOOLTIP_SORT_ORDER),
 				NWidget(WWT_DROPDOWN, COLOUR_BROWN, WID_TD_SORT_CRITERIA), SetDataTip(STR_JUST_STRING, STR_TOOLTIP_SORT_CRITERIA),
-				NWidget(WWT_PANEL, COLOUR_BROWN), SetResize(1, 0), EndContainer(),
+				NWidget(WWT_EDITBOX, COLOUR_BROWN, WID_SL_FILTER), SetFill(35, 12),
+						SetDataTip(STR_LIST_FILTER_OSKTITLE, STR_LIST_FILTER_TOOLTIP),
```
Comment (nielsmh) [inline src/town_gui.cpp]: Breaking the line here doesn't really match the surrounding code. For the most part we don't break long lines, unless it really improves readability.

File: src/town_gui.cpp [hunk 3]
```diff
@@ -644,6 +646,10 @@ struct TownDirectoryWindow : public Window {
 	static const StringID sorter_names[];
 	static GUITownList::SortFunction * const sorter_funcs[];
 
+	StringFilter string_filter; ///< Filter for towns
+	QueryString townname_editbox; ///< Filter editbox;
+	SmallVector<bool, 32> town_items_shown; ///< Map of the filtered out town names
```
Comment (nielsmh) [inline src/town_gui.cpp:650]: You should align the comments here, so they all start on the same column.

File: src/town_gui.cpp [hunk 4]
```diff
@@ -922,16 +931,45 @@ struct TownDirectoryWindow : public Window {
...
+	}
+
 	/**
 	 * Some data on this window has become invalid.
 	 * @param data Information about the changed data.
 	 * @param gui_scope Whether the call is done from GUI scope. You may not do everything when not in GUI scope. See #InvalidateWindowData() for details.
 	 */
 	virtual void OnInvalidateData(int data = 0, bool gui_scope = true)
 	{
+		char buf[64];
```
Comment (nielsmh) [inline src/town_gui.cpp]: Magical number, I think this should be `MAX_LENGTH_TOWN_NAME_CHARS` instead.

**Review by nielsmh — COMMENTED** [review]
Thanks for your contribution! Overall it looks good, there's a few things you should change, mostly cosmetic.

You also need to change the commit message to follow the scheme used, always starting with a keyword (like Add, Change, Fix), a colon, and then the message. Do this with a commit amend in Git, and force-push your branch. (Amending and force-pushing is often frowned upon, but necessary in this case. Github handles it correctly and updates the PR.)

**Comment by telk5093** [comment]
@nielsmh Thanks for your kind comments. I've corrected, is it okay?

**Comment by nielsmh** [comment]
I don't have anything more for now. Will find time to test it later :)

**Comment by J0anJosep** [comment]
Thanks for the PR.
Each time a new widget is added, run /src/script/api/generate_widget.sh and add the modified lines into your commit.

**Comment by telk5093** [comment]
@J0anJosep Thanks for your notice, I committed, is it right?

**Comment by LordAro** [comment]
You'll want to rebase & force-push to merge the commits together, but that's the right idea

**Comment by planetmaker** [comment]
It looks nice. However the sorting doesn't seem to be stable even when I do not change input into the search box (anymore). The resulting list flickers and changes order randomly for me. That doesn't happen with the full list where the entries immediately follow eachother.

I currently can only speculate as to the reasons: I assume a sub-list is generated from the whole list - and display sometimes happens before sorting, but after generation of the sub-list - but most times it happens correctly after sub-list generation and its sorting.

Looks to me you already consider only re-generating the list, when the input field actually changed, though.

Honestly I'm not sure whether this, my observation, is beyond the scope of this patch which just makes use of the existing sort framework. I lean toward "out-of-scope" - yet this makes a bug or race condition visible.

**Review by LordAro — CHANGES_REQUESTED** [review]
As per @planetmaker's comments

**Comment by telk5093** [comment]
@planetmaker Thanks for your comment, even though I cannot understand 100% what you're meant to because of my lack of english/coding skills.
I'll try to rewrite my code. (of course I think it will take a long time ;p)

**Pull request closed without merging**